### ISSUE TITLE
systemd-vconsole-setup: Failed to lock /dev/console: Input/output error

### ISSUE THREAD

**Issue by moviuro** [systemd/systemd#30501]
### systemd version the issue has been seen with

255-2

### Used distribution

archlinux

### Linux kernel version used

6.1.68-1-lts

### CPU architectures issue was seen on

x86_64

### Component

other

### Expected behaviour you didn't see

* Keymap is applied during boot so I can type my LUKS passphrase with my local keyboard mapping to the print on the keyboard (typing "A" gives "A", not "Q").
* Keymap is applied on ttys, so I can type my username and password with my local keyboard mapping to the print on the keyboard

### Unexpected behaviour you saw

* Keymap stays QWERTY during early boot (LUKS)
* Keymap stays QWERTY after boot (login on tty)
* This is new behavior. It didn't occur recently on other machines

The journal doesn't show anything DEBUG even though I have `Environment=SYSTEMD_LOG_LEVEL=debug`?

```
# journalctl -u systemd-vconsole-setup.service -b
Dec 16 08:58:42 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 08:58:42 archlinux systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
Dec 16 08:58:42 archlinux systemd[1]: systemd-vconsole-setup.service: Failed with result 'exit-code'.
Dec 16 08:58:42 archlinux systemd[1]: Failed to start Virtual Console Setup.
Dec 16 08:58:42 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 08:58:42 archlinux systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
Dec 16 08:58:42 archlinux systemd[1]: systemd-vconsole-setup.service: Failed with result 'exit-code'.
Dec 16 08:58:42 archlinux systemd[1]: Failed to start Virtual Console Setup.
Dec 16 08:58:45 toxoplasmosis systemd[1]: Starting Virtual Console Setup...
Dec 16 08:58:45 toxoplasmosis systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
Dec 16 08:58:45 toxoplasmosis systemd[1]: systemd-vconsole-setup.service: Failed with result 'exit-code'.
Dec 16 08:58:45 toxoplasmosis systemd[1]: Failed to start Virtual Console Setup.
Dec 16 08:58:52 toxoplasmosis systemd[1]: Starting Virtual Console Setup...
Dec 16 08:58:52 toxoplasmosis systemd-vconsole-setup[1076]: Failed to lock /dev/console: Input/output error
Dec 16 08:58:52 toxoplasmosis systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
Dec 16 08:58:52 toxoplasmosis systemd[1]: systemd-vconsole-setup.service: Failed with result 'exit-code'.
Dec 16 08:58:52 toxoplasmosis systemd[1]: Failed to start Virtual Console Setup.
```

### Steps to reproduce the problem

```ini
# systemctl cat systemd-vconsole-setup.service
# /usr/lib/systemd/system/systemd-vconsole-setup.service
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Virtual Console Setup
Documentation=man:systemd-vconsole-setup.service(8) man:vconsole.conf(5)

ConditionPathExists=/dev/tty0

DefaultDependencies=no
Before=sysinit.target
Before=initrd-switch-root.target shutdown.target

# This unit will be restarted by udev whenever a new vtcon device appears or is
# triggered. Usually that happens just a handful of times during boot, but some
# systems may have hundreds or thousands of serial consoles connected, so let's
# just disable the limit altogether.
StartLimitIntervalSec=0

[Service]
Type=oneshot
# This service will be restarted by udev whenever a new vtcon device appears.
# If the previous instance is still running, it shall be interrupted without
# error.
SuccessExitStatus=SIGTERM
RemainAfterExit=yes

ExecStart=/usr/lib/systemd/systemd-vconsole-setup

ImportCredential=vconsole.*

# /etc/systemd/system/systemd-vconsole-setup.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```

```sh
# cat /etc/vconsole.conf 
# Written by systemd-localed(8) or systemd-firstboot(1), read by systemd-localed
# and systemd-vconsole-setup(8). Use localectl(1) to update this file.
KEYMAP=fr
XKBLAYOUT=fr
XKBMODEL=pc105
XKBOPTIONS=terminate:ctrl_alt_bksp
```

```
# localectl  
System Locale: LANG=en_US.UTF-8
               LC_NUMERIC=en_DK.UTF-8
               LC_TIME=en_DK.UTF-8
               LC_COLLATE=C
               LC_MONETARY=fr_FR.UTF-8
               LC_NAME=fr_FR.UTF-8
               LC_ADDRESS=fr_FR.UTF-8
               LC_TELEPHONE=fr_FR.UTF-8
               LC_MEASUREMENT=fr_FR.UTF-8
    VC Keymap: fr
   X11 Layout: fr
    X11 Model: pc105
  X11 Options: terminate:ctrl_alt_bksp
```

`mkinicpio.conf(5)` has been in use **as-is** since 2021.

```sh
# grep -vE '^($|#)' /etc/mkinitcpio.conf
MODULES=(vfat)
BINARIES=()
FILES=()
HOOKS=(systemd autodetect modconf block sd-vconsole keyboard sd-encrypt filesystems)
```

```
# systemctl --version                         
systemd 255 (255-2-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

----

My other machine that has a pretty similar setup doesn't exhibit any symptom. Keymap is correctly set to French and I can happily type away (LUKS, login).

```sh
# cat /etc/vconsole.conf # -rw-r--r-- 1 root root 13 Jan 18  2016 /etc/vconsole.conf
KEYMAP=fr-pc
```

```
# journalctl -u systemd-vconsole-setup -b
Dec 16 09:10:14 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Finished Virtual Console Setup.
Dec 16 09:10:14 archlinux systemd[1]: systemd-vconsole-setup.service: Deactivated successfully.
Dec 16 09:10:14 archlinux systemd[1]: Stopped Virtual Console Setup.
Dec 16 09:10:14 archlinux systemd[1]: Stopping Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Finished Virtual Console Setup.
Dec 16 09:10:27 archlinux systemd[1]: systemd-vconsole-setup.service: Deactivated successfully.
Dec 16 09:10:27 archlinux systemd[1]: Stopped Virtual Console Setup.
Dec 16 09:10:14 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Finished Virtual Console Setup.
Dec 16 09:10:14 archlinux systemd[1]: systemd-vconsole-setup.service: Deactivated successfully.
Dec 16 09:10:14 archlinux systemd[1]: Stopped Virtual Console Setup.
Dec 16 09:10:14 archlinux systemd[1]: Stopping Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:14 archlinux systemd[1]: Finished Virtual Console Setup.
Dec 16 09:10:27 archlinux systemd[1]: systemd-vconsole-setup.service: Deactivated successfully.
Dec 16 09:10:27 archlinux systemd[1]: Stopped Virtual Console Setup.
Dec 16 09:10:30 psychoticdelirium systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:30 psychoticdelirium systemd[1]: systemd-vconsole-setup.service: Deactivated successfully.
Dec 16 09:10:30 psychoticdelirium systemd[1]: Stopped Virtual Console Setup.
Dec 16 09:10:30 psychoticdelirium systemd[1]: Starting Virtual Console Setup...
Dec 16 09:10:31 psychoticdelirium systemd[1]: Finished Virtual Console Setup.
```

```sh
# grep -vE '^($|#)' /etc/mkinitcpio.conf # -rw-r--r-- 1 root root 2607 Jun  5  2021 /etc/mkinitcpio.conf
MODULES=()
BINARIES=()
FILES=()
HOOKS=(systemd autodetect modconf block sd-vconsole keyboard sd-encrypt filesystems)
```

```
# localectl 
System Locale: LANG=en_US.UTF-8
               LC_TIME=en_DK.UTF-8
               LC_MONETARY=fr_FR.UTF-8
               LC_NAME=fr_FR.UTF-8
               LC_ADDRESS=fr_FR.UTF-8
               LC_TELEPHONE=fr_FR.UTF-8
               LC_MEASUREMENT=fr_FR.UTF-8
    VC Keymap: fr-pc
   X11 Layout: fr
```

```
# systemctl --version
systemd 255 (255-2-arch)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

**Comment by moviuro** [comment]
This is still an issue on:

* systemd 255 (255.3-1.1-arch) `+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified`
* Linux toxoplasmosis 6.6.14-1.1-lts #1 SMP PREEMPT_DYNAMIC Sat, 27 Jan 2024 01:57:42 +0000 x86_64 GNU/Linux

**Comment by sedlund** [comment]
255.2

```console
Feb 25 14:14:36 sgp1 systemd-vconsole-setup[635762]: Failed to lock /dev/console: Input/output error
Feb 25 14:14:36 sgp1 systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
```

causes my deploys to fail.  any ideas?

**Comment by CRCinAU** [comment]
This also happens on the upcoming Fedora 40:

```
× systemd-vconsole-setup.service - Virtual Console Setup
     Loaded: loaded (/usr/lib/systemd/system/systemd-vconsole-setup.service; static)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Mon 2024-04-08 13:49:30 UTC; 55s ago
       Docs: man:systemd-vconsole-setup.service(8)
             man:vconsole.conf(5)
    Process: 722 ExecStart=/usr/lib/systemd/systemd-vconsole-setup (code=exited, status=1/FAILURE)
   Main PID: 722 (code=exited, status=1/FAILURE)
        CPU: 6ms

Apr 08 13:49:30 systemd[1]: Starting systemd-vconsole-setup.service - Virtual Console Setup...
Apr 08 13:49:30 systemd-vconsole-setup[722]: Failed to lock /dev/console: Input/output error
Apr 08 13:49:30 systemd[1]: systemd-vconsole-setup.service: Main process exited, code=exited, status=1/FAILURE
Apr 08 13:49:30 systemd[1]: systemd-vconsole-setup.service: Failed with result 'exit-code'.
Apr 08 13:49:30 systemd[1]: Failed to start systemd-vconsole-setup.service - Virtual Console Setup.
```

Package versions:
```
$ rpm -qa | grep systemd | sort
python3-systemd-235-9.fc40.x86_64
rpm-plugin-systemd-inhibit-4.19.1.1-1.fc40.x86_64
systemd-255.4-1.fc40.x86_64
systemd-boot-unsigned-255.4-1.fc40.x86_64
systemd-libs-255.4-1.fc40.x86_64
systemd-networkd-255.4-1.fc40.x86_64
systemd-oomd-defaults-255.4-1.fc40.noarch
systemd-pam-255.4-1.fc40.x86_64
systemd-resolved-255.4-1.fc40.x86_64
systemd-udev-255.4-1.fc40.x86_64
```

Fedora bug report:
https://bugzilla.redhat.com/show_bug.cgi?id=2273069

**Comment by JM1** [comment]
What is your `/proc/cmdline`? Does it contain any `console=` entries? What happens if you remove these `console=` entries?

**Comment by CRCinAU** [comment]
```
initrd=\1f3beac22b174372878b369646feb3f3\6.8.5-301.fc40.x86_64\initrd root=UUID=a280b604-6023-4ba5-bb9e-80d612f84b0d ro rootflags=subvol=root no_timer_check net.ifnames=0 console=tty1 console=ttyS0,115200n8 audit=0 selinux=0 systemd.machine_id=1f3beac22b174372878b369646feb3f3
```

I removed both console= entries, and now I can't seem to log into the machine - via the local console, or via SSH...

**Comment by JM1** [comment]
Could you post your QEMU command with args?

You could also add `systemd.mask=systemd-vconsole-setup.service` to your kernel cmdline (while keeping `console=`) to complete booting, and then use something like `TERM=linux setterm --blank 0 > /dev/tty0` etc. to find the offending tty.

**Comment by JM1** [comment]
FYI [vconsole-setup.c](https://github.com/systemd/systemd/commits/main/src/vconsole/vconsole-setup.c) got a couple of fixes lately, would be interesting to know if any of that fixed your issue.

**Comment by JM1** [comment]
FYI You could [pass the correct tty to `systemd-vconsole-setup` as a cmd arg](https://github.com/systemd/systemd/blob/db11bab38ccf1ed257f310d29070843d4c58ea01/src/vconsole/vconsole-setup.c#L596). For example, run `systemctl edit systemd-vconsole-setup.service` and paste something similar to:

```
[Service]
ExecStart=
ExecStart=/usr/lib/systemd/systemd-vconsole-setup /dev/tty1
```

Adapt `/dev/tty1` to a tty which did not fail when using it with `setterm` (see above).

**Comment by CRCinAU** [comment]
I've been tinkering with this a bit more.... I restored the previous console= command options, with no real success...

If I have the i915 passthru enabled, I can't login via the VirtIO-GPU, SSH, or a serial console... As soon as I remove the i915 passthru, everything works as normal again.

I'm not exactly sure on how to proceed with this...

**Comment by JM1** [comment]
What does "can't login" mean? Please add kernel arg `systemd.mask=systemd-vconsole-setup.service` and try again. If you are still not able to log in, not even with SSH, there are probably more failing services than just `systemd-vconsole-setup.service`. With `console=` and `earlyprintk=` both pointing to a serial console of your VM, you do not see any reports from systemd about failing services?

**Comment by CRCinAU** [comment]
By "can't login" - I mean I get no console login prompt via serial console, on the virtio GPU interface, or successful login via SSH.

The SSH port answers (and a telnet will show the SSH banner), but the login never completes.

All the startup applications work, docker services start etc etc just fine.

I haven't tried any of the other systemd parts you mentioned as yet - I'll need to sit down with a bit of time and explore some more.

**Comment by Mithi83** [comment]
I'm the person who wrote the initial Fedora bug report in https://bugzilla.redhat.com/show_bug.cgi?id=2273069

Thanks for your discussion, which led me to discover a way (or rather two ways) to fix my personal problem.

My setup has a bit of history: The VM is more than a decade old and therefore probably has some legacy to carry around. In particular the VM used to run on a Fedora host under libvirt. To use "virsh console" at some point I decided to change the grub cmdline to include "console=tty console=ttyS0". On that Fedora host qemu provided some sort of virtual serial device to the VM that "virsh console" could use. Fast forward a few years: That Fedora host was replaced with a Proxmox server, the VM image was transferred to that server. The hardware description of the VM however was created anew and since Proxmox can use the VGA Output easily I had no need to add a virtual serial device to the VM hardware description and forgot to do that. The grub cmdline still contained the "console=ttyS0" part though, which was a mismatch and my own mistake. Funny enough: Up until Fedora 39 (including) it still worked despite that mismatch, whereas under Fedora 40 it ran into the problem.

Since I read your comments I found two solutions to my problem. Either one of them is sufficient to have a working setup again:
1. Remove "console=ttyS0" from the grub cmdline
2. Add a virtual serial device to the hardware description of the VM in Proxmox

My guess is that a non-existing (maybe also non-responding?) device may cause that problem. Also I'm not sure what part of the chain kernel...systemd is responsible, since I don't know how the console= cmdline variable is used in the kernel and/or systemd. Both would have changed between Fedora 39 and 40 I presume.
